OpenRA.Mods.Common/Traits/Attack/AttackBase.cs

```diff
@@ -211,12 +208,21 @@ public bool HasAnyValidWeapons(Target t)
 
 			// PERF: Avoid LINQ.
 			foreach (var armament in Armaments)
-				if (!armament.OutOfAmmo && armament.Weapon.IsValidAgainst(t, self.World, self))
+			{
+				// If 'onlyTargetCenterWeapons' is true, we don't care about current ammo state, only Weapon.TargetActorCenter
+				var checkIsValidAgainst = onlyTargetCenterWeapons ? armament.Weapon.TargetActorCenter : !armament.OutOfAmmo;
+				if (checkIsValidAgainst && armament.Weapon.IsValidAgainst(t, self.World, self))
 					return true;
+			}
 
 			return false;
 		}
 
+		public virtual WPos GetTargetPosition(WPos pos, Target target, bool onlyTargetCenterWeapons = false)
+		{
+			return HasAnyValidWeapons(target, onlyTargetCenterWeapons) ? target.CenterPosition : target.Positions.PositionClosestTo(pos);
```

Comment:
HasAnyValidWeapons with that new arg is fine for me, my comment was about WPos GetTargetPosition function. It will return CenterPosition regardless if onlyTargetCenterWeapons  is true or false.